pkg/maintenance/minio.go

```diff
@@ -0,0 +1,154 @@
+package maintenance
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"regexp"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/spf13/viper"
+	"github.com/vshn/appcat/v4/pkg/maintenance/helm"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	minioURL        = "https://hub.docker.com/v2/repositories/minio/minio/tags/?page_size=100"
+	dateRegex       = minioTagPrefix + `(?P<date>.*Z$)`
+	minioTimeLayout = "2006-01-02T15-04-05Z07:00"
+	minioTagPrefix  = "RELEASE."
+	earliestTag     = "2023-09-20T22-49-55Z"
+)
+
+// Minio contains all necessary dependencies to successfully run a minio maintenance
+type Minio struct {
+	k8sClient  client.Client
+	httpClient *http.Client
+	log        logr.Logger
+}
+
+// NewMinio returns a new Minio object
+func NewMinio(c client.Client, hc *http.Client) Minio {
+	return Minio{
+		k8sClient:  c,
+		httpClient: hc,
+	}
+}
+
+// DoMaintenance will run minios's maintenance script.
+func (m *Minio) DoMaintenance(ctx context.Context) error {
+	m.log = logr.FromContextOrDiscard(ctx).WithValues("type", "minio")
+	patcher := helm.NewImagePatcher(m.k8sClient, m.httpClient, m.log)
+
+	valuesPath := helm.NewValuePath("image", "tag")
+
+	err := m.ensureTagIsNotNil(ctx, valuesPath)
+	if err != nil {
+		return err
+	}
+
+	return patcher.DoMaintenance(ctx, minioURL, valuesPath, compareMinioVersions)
+}
+
+// compareMinioVersions specifically checks for new Minio versions
+// Minio has a pretty unique rolling release cycle which is based on the date.
+// Each Minio image tag contains the date when it was created.
+// The form of the tags is like this: `RELEASE.2023-09-16T01-01-47Z`
+func compareMinioVersions(results []helm.Result, currentTag string) (string, error) {
+
+	if currentTag == "" {
+		currentTag = minioTagPrefix + earliestTag
```

Comment:
Any reason you have to deduce that? I mean you should be able to get the version of the current release no?

Reply:
We don't set it initially, so the release takes the defaults, but that is not actually populated in the helm values. This logic is to catch this case, which is the very first maintenance after provisioning.